src/Servers.jl

```diff
@@ -130,278 +109,196 @@ function check_rate_limit(tcp;
+Base.isopen(s::Server2) = isopen(s.server)
+Base.close(s::Server2) = close(s.server)
+
+Sockets.accept(s::Server2{Nothing, S}) where {S} = Sockets.accept(s.server)
+Sockets.accept(s::Server2) = getsslcontext(accept(s.server), s.ssl)
 
 function getsslcontext(tcp, sslconfig)
-    ssl = SSLContext()
-    setup!(ssl, sslconfig)
-    associate!(ssl, tcp)
-    handshake!(ssl)
+    ssl = MbedTLS.SSLContext()
+    MbedTLS.setup!(ssl, sslconfig)
+    MbedTLS.associate!(ssl, tcp)
+    MbedTLS.handshake!(ssl)
     return ssl
 end
-getsslcontext(tcp, ::Nothing) = tcp
-
-const nolimit = typemax(Int)
 
 """
-    HTTP.listen([host=Sockets.localhost[, port=8081]]; <keyword arguments>) do http
+    HTTP.listen([host=Sockets.localhost[, port=8081]]; kw...) do req
         ...
     end
+    HTTP.listen(handler::HTTP.Handler, host=Sockets.localhost, port=8081; kw...)
 
-Listen for HTTP connections and execute the `do` function for each request.
+Listen for HTTP connections and either execute the `do` function for each request, or dispatch to the
+provided `handler`. Both the function or `handler` can accept an `HTTP.Request` object, or a raw
+`HTTP.Stream` connection to read & write from directly; for the latter, pass the `stream=true`
+keyword argument to operate on the `HTTP.Stream` directly.
 
 Optional keyword arguments:
- - `ssl::Bool = false`, use https.
- - `require_ssl_verification = true`, pass `MBEDTLS_SSL_VERIFY_REQUIRED` to
-   the mbed TLS library.
-   ["... peer must present a valid certificate, handshake is aborted if
-     verification failed."](https://tls.mbed.org/api/ssl_8h.html#a5695285c9dbfefec295012b566290f37)
- - `sslconfig = SSLConfig(require_ssl_verification)`
- - `pipeline_limit = 16`, number of concurrent requests per connection.
+ - `sslconfig=nothing`: Provide an `MbedTLS.SSLConfig` object to handle ssl connections
  - `reuse_limit = nolimit`, number of times a connection is allowed to be reused
                             after the first request.
  - `tcpisvalid::Function (::TCPSocket) -> Bool`, check accepted connection before
     processing requests. e.g. to implement source IP filtering, rate-limiting,
     etc.
+ - `readtimeout::Int=0`: # of seconds to wait on an incoming request before closing a connection
+ - `reuseaddr::Bool=false`: whether multiple servers should be allowed to listen on the same port
  - `tcpref::Ref{Base.IOServer}`, this reference is set to the underlying
                                  `IOServer`. e.g. to allow closing the server.
+ - `connectioncounter::Ref{Int}`: a `Ref{Int}` that can be used to track the # of currently open (i.e
+        currently being handled) connections for a server
+ - `ratelimit`: a `Rational{Int}` of the form `5//1` indicating how many `messages//second`
+        should be allowed per client IP address; requests exceeding the rate limit will be auto-closed
+ - `stream::Bool=false`: whether the handler should operate on an `HTTP.Stream` to read & write directly
```

Comment:
Suggest: "pass a `HTTP.Stream` to the handler function (instead of a `HTTP.Request`).